cargo-near/src/abi/generation.rs

```diff
@@ -0,0 +1,165 @@
+    let dylib_dir = dylib_path.parent().ok_or_else(|| {
+        anyhow::anyhow!(
+            "Unable to infer the directory containing dylib file: {}",
+            dylib_path.display()
+        )
+    })?;
+    let dylib_name = dylib_path
+        .file_stem()
+        .ok_or_else(|| anyhow::anyhow!("Generated dylib is not a file: {}", dylib_path.display()))?
+        .to_str()
+        .ok_or_else(|| {
+            anyhow::anyhow!(
+                "Unable to infer the directory containing dylib file: {}",
+                dylib_path.display()
+            )
+        })?;
+
+    let dylib_name = if let Some(dylib_name_stripped) = dylib_name.strip_prefix("lib") {
+        dylib_name_stripped
+    } else {
+        anyhow::bail!(
+            "Expected the generated dylib file to start with 'lib', but got '{}'",
+            dylib_name
+        );
+    };
+
+    let cargo_link_lib = format!("cargo:rustc-link-lib=dylib={}", &dylib_name);
+    let cargo_link_search = format!("cargo:rustc-link-search=all={}", dylib_dir.display());
+
+    let build_rs = quote! {
+        fn main() {
+            println!(#cargo_link_lib);
+            println!(#cargo_link_search);
+        }
+    }
+    .to_string();
+    let build_rs_file = syn::parse_file(&build_rs).unwrap();
+    let build_rs = prettyplease::unparse(&build_rs_file);
+
+    log::debug!("build.rs contents:\n{}", &build_rs);
+
+    Ok(build_rs)
+}
+
+pub(crate) fn generate_main_rs(dylib_path: &Path) -> anyhow::Result<String> {
+    let dylib_file_contents = fs::read(&dylib_path)?;
+    let object = symbolic_debuginfo::Object::parse(&dylib_file_contents)?;
+    log::debug!(
+        "A dylib was built at {:?} with format {} for architecture {}",
+        &dylib_path,
+        &object.file_format(),
+        &object.arch()
+    );
+    let near_abi_symbols = object
+        .symbols()
+        .flat_map(|sym| sym.name)
+        .filter(|sym_name| sym_name.starts_with("__near_abi_"))
+        .map(|sym_name| sym_name.to_string())
+        .collect::<HashSet<_>>();
+    if near_abi_symbols.is_empty() {
+        anyhow::bail!("No NEAR ABI symbols found in the dylib");
+    }
+    log::debug!("Detected NEAR ABI symbols: {:?}", &near_abi_symbols);
+
+    let near_abi_function_defs = near_abi_symbols.iter().map(|s| {
+        let name = format_ident!("{}", s);
+        quote! {
+            fn #name() -> near_sdk::__private::ChunkedAbiEntry;
+        }
+    });
+    let near_abi_function_invocations = near_abi_symbols.iter().map(|s| {
+        let name = format_ident!("{}", s);
+        quote! {
+            unsafe { #name() }
+        }
+    });
+
+    let main_rs = quote! {
+        extern "Rust" {
+            #(#near_abi_function_defs)*
+        }
+
+        fn main() -> Result<(), std::io::Error> {
+            let abi_entries: Vec<near_sdk::__private::ChunkedAbiEntry> = vec![#(#near_abi_function_invocations),*];
+            let contents = serde_json::to_string_pretty(&abi_entries)?;
```

Comment:
We can actually move this bit to the SDK, simplifying things. See alt: https://github.com/near/cargo-near/pull/74

Reply:
Not really, it may look similar but there is an important difference that I have touched upon here: https://github.com/near/cargo-near/pull/74#pullrequestreview-1147284614

Reply:
Yeah, I was thinking a Rust `String` might be safe, but since we don't have that guarantee, it's just best to switch to the C ABI.